lib/elrequest.js

```diff
@@ -0,0 +1,187 @@
+    // socket is *created* ("socket" event) and ending when the socket
+    // is connected.
+    req.on('socket', function (socket) {
+      // log.trace({url: url}, 'start connectTimeout')
+      var connectTimer = setTimeout(function onConnectTimeout () {
+        // log.trace({url: url}, 'connectTimeout')
+        req.emit('connectTimeout')
+      }, connectTimeout)
+
+      socket.on('connect', function () {
+        // log.trace({url: url}, 'socket connected, clear connectTimeout')
+        clearTimeout(connectTimer)
+        connectTimer = null
+      })
+      socket.on('close', function () {
+        if (connectTimer) {
+          // log.trace({url: url}, 'socket close with active connectTimer, clear connectTimeout')
+          clearTimeout(connectTimer)
+        }
+      })
+    })
+  }
+
+  return req
+}
+
+module.exports = {
+  elRequest: elRequest
+}
+
+// ---- mainline
+
+// This main is only intended to demonstrate usage of this lib; not to be a
+// useful tool.
+//
+// Example:
+//      node elrequest.js https://www.elastic.co 100 1000 > elastic.html
+//
+// Example: download a ~1GB file, requiring a 30ms connect time, 2s idle timeout
+//      node elrequest.js http://mirror.uoregon.edu/ubuntu-releases/20.04/ubuntu-20.04.1-live-server-amd64.iso 30 2000 > ubuntu.iso
+//
+// Example: connect timeout because google port 81 drops TCP SYN packets:
+//      node elrequest.js http://www.google.com:81/foo 1000 10000
+//
+// Example: use `NODE_DEBUG=*` to see internal node debugging details
+//      NODE_DEBUG=* node elrequest.js $url 30 1000
+//
+// Example: quick connection, slow response
+//      % cat server.js
+//      var http = require('http');
+//      http.createServer(function (req, res) {
+//          console.log('SERVER: got request')
+//          setTimeout((function() {
+//              res.writeHead(200, {'Content-Type': 'text/plain'});
+//              res.write('line one\n')
+//              setTimeout((function() {
+//                  res.write('line two\n')
+//                  setTimeout((function() {
+//                      res.write('line three\n')
+//                      res.end()
+//                      console.log('SERVER: responded')
+//                  }), 500);
+//              }), 500);
+//          }), 500);
+//      }).listen(8080);
+//      % node server.js &
+//      % node elrequest.js http://127.0.0.1:8080/ 10 1000
+function main (argv) {
+  if (argv.length !== 5) {
+    process.stderr.write('elrequest: error: incorrect number of args\n')
+    process.stderr.write('usage: elrequest $url $connectTimeoutMs $timeoutMs\n')
+    process.exitCode = 1
+    return
+  }
+  const url = argv[2]
+  const connectTimeout = Number(argv[3])
+  const timeout = Number(argv[4])
+
+  var req = elRequest(url, {
+    timeout: timeout,
+    connectTimeout: connectTimeout
+    // TODO: log support
+  }, function onRes (res) {
+    res.pipe(process.stdout)
+  })
+
+  req.on('timeout', function () {
```

Comment:
Question: Is listening for the timeout event required in Node.js -- or will the request object can/destroy itself when the timeout is reached?

Reply:
https://nodejs.org/api/all.html#http_event_timeout 

> Emitted when the underlying socket times out from inactivity. This only notifies that the socket has been idle. The request must be aborted manually.

So, if one passes in a `options.timeout` for the request, then one must manually `req.destroy()`. 

Also note: https://nodejs.org/api/all.html#http_request_abort   hence why I'm using `req.destroy()` rather than `req.abort()`.